theano/tensor/nnet/corr.py

```diff
@@ -75,6 +75,34 @@ def __str__(self):
             str(self.subsample),
             str(self.filter_dilation))
 
+    def cast(self, in1, in2):
+        """
+        Upcast input variables if neccesary.
+        """
+        float_types = ['float32', 'float64']
+        dtype_1 = in1.type.dtype
+        dtype_2 = in2.type.dtype
+
+        assert dtype_1 in float_types
+        assert dtype_2 in float_types
+
+        if dtype_1 == 'float64' or dtype_2 == 'float64':
+            dtype_o = 'float64'
+        else:
+            dtype_o = 'float32'
+
+        if dtype_1 != dtype_o:
+            out1 = in1.astype(dtype_o)
+        else:
+            out1 = in1
+
+        if dtype_2 != dtype_o:
+            out2 = in2.astype(dtype_o)
+        else:
+            out2 = in2
+
+        return out1, out2
+
```

Comment:
I think this could be reduced to:
```python
def cast(self, in1, in2):
    dtype = theano.scalar.upcast(in1.dtype, in2.dtype)
    return in1.astype(dtype), in2.astype(dtype)
```
`astype` calls `T.cast`, which returns the variable unchanged if it's the correct type (in contrast to numpy, where `.astype` always creates a copy).